edge/pkg/devicetwin/dtmanager/communicate_test.go

```diff
@@ -0,0 +1,319 @@
+		t.Run(test.name, func(t *testing.T) {
+			got, err := dealSendToCloud(test.context, test.resource, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("dealSendToCloud() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("dealSendToCloud() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestDealLifeCycle is function to test dealLifeCycle().
+func TestDealLifeCycle(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContext, _ := dtcontext.InitDTContext(mainContext)
+	tests := []struct {
+		name     string
+		context  *dtcontext.DTContext
+		resource string
+		msg      interface{}
+		want     interface{}
+		wantErr  error
+	}{
+		{
+			name:     "dealLifeCycleTest-WrongMessageFormat",
+			context:  dtContext,
+			resource: "",
+			msg:      "",
+			want:     nil,
+			wantErr:  errors.New("msg not Message type"),
+		},
+		{
+			name:     "dealLifeCycleTest-CloudConnected",
+			context:  dtContext,
+			resource: "",
+			msg:      &model.Message{Content: model.CloudConnected},
+			want:     nil,
+			wantErr:  nil,
+		},
+		{
+			name:     "dealLifeCycleTest-CloudNotConnected",
+			context:  dtContext,
+			resource: "",
+			msg:      &model.Message{Content: model.CloudDisconnected},
+			want:     nil,
+			wantErr:  nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := dealLifeCycle(test.context, test.resource, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("dealLifeCycle() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("dealLifeCycle() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestDealConfirm is function to test dealConfirm().
+func TestDealConfirm(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContext, _ := dtcontext.InitDTContext(mainContext)
+	tests := []struct {
+		name     string
+		context  *dtcontext.DTContext
+		resource string
+		msg      interface{}
+		want     interface{}
+		wantErr  error
+	}{
+		{
+			name:    "dealConfirmTest-WrongMsg",
+			context: dtContext,
+			msg:     "",
+			want:    nil,
+			wantErr: errors.New("CommModule deal confirm, type not correct"),
+		},
+		{
+			name:    "dealConfirmTest-CorrectMsg",
+			context: dtContext,
+			msg:     &model.Message{},
+			want:    nil,
+			wantErr: nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := dealConfirm(test.context, test.resource, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("dealConfirm() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("dealConfirm() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestDetailRequest is function to test detailRequest().
+func TestDetailRequest(t *testing.T) {
```

Comment:
why do we have a separate test case for testing the detailRequest method ? This is called as a part of dealLifecycle method. the test for dealLifecycle should have a case where this method is called.